src/high-level/constructors.lisp

```diff
@@ -0,0 +1,108 @@
+;;;; constructors.lisp
+;;;;
+;;;; Author: Cole Scott
+
+(in-package #:magicl)
+
+(defconstant +default-tensor-type+ 'double-float)
+
+(defgeneric empty (shape &key type order)
+  (:documentation "Create an empty tensor")
+  (:method (shape &key (type +default-tensor-type+) order)
+    (check-type shape shape)
+    (let ((tensor-type (compatible-tensor-constructors type)))
+      (specialize-tensor (make-tensor shape tensor-type type :order order)))))
+
+(defgeneric const (const shape &key type order)
+  (:documentation "Create tensor with all elements equal to a constant value")
+  (:method (const shape &key type order)
+    (check-type shape shape)
+    (let ((tensor-class
+            (if (null type)
+                (compatible-tensor-constructors-from-value const)
+                (compatible-tensor-constructors type)))
+          (element-type
+            (if (null type)
+                (upgraded-array-element-type (type-of const))
+                type)))
+      (specialize-tensor (make-tensor shape tensor-class element-type :order order :initial-element (coerce const element-type))))))
+
+(defgeneric rand (min max shape &key type distribution)
+  (:documentation "Create tensor with elements random in the range [0,limit]")
```

Comment:
Maybe `distribution` should be a function, defaulting to `random`.

Reply:
If you want SBCL to inline the `distribution` (e.g. given as a lambda), IMO going to the generic function is not a good idea...